vpr/src/base/vpr_constraints_writer.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * vpr_constraints_writer.cpp
+ *
+ *      Author: khalid88
+ */
+
+#include "vpr_constraints_serializer.h"
+#include "vpr_constraints_uxsdcxx.h"
+
+#include "vtr_time.h"
+
+#include "globals.h"
+#include "pugixml.hpp"
+#include "pugixml_util.hpp"
+#include "echo_files.h"
+
+#include <fstream>
+#include "vpr_constraints_writer.h"
+
+void write_vpr_floorplan_constraints(const char* file_name, int expand, int subtile) {
+    auto& place_ctx = g_vpr_ctx.placement();
+    auto& atom_ctx = g_vpr_ctx.atom();
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+
+    //Fill in all the info needed for locking atoms to their locations
+    VprConstraints constraints;
+
+    std::map<ClusterBlockId, std::vector<AtomBlockId>> cluster_atoms;
+
+    for (auto blk_id : cluster_ctx.clb_nlist.blocks()) {
+        cluster_atoms.insert({blk_id, std::vector<AtomBlockId>()});
+    }
+
+    for (auto atom_blk_id : atom_ctx.nlist.blocks()) {
+        ClusterBlockId clb_index = atom_ctx.lookup.atom_clb(atom_blk_id);
```

Comment:
this lookup may be looking into a vector that is just like the map I'm using, take a look

Reply:
make sure there is no data structure that is already doing this, if there is, make a utility function that does this on call. the utility function can go in with some kind of clb netlist utils file,

Reply:
also look at clb netlist to see if it has this

Reply:
can add to clustered_netlist_utils if you add the utility function

Reply:
can also make a small class whose constructor creates the vector as a private member, and some function to say give me the atoms in this cluster, can add this class to clustered_netlist_utils.cpp

Reply:
for clusters_echo, will need to use this vector as well, so go change that code in cluster.cpp